src/mgmt/tests/srv_drpc_tests.c

```diff
@@ -786,13 +786,205 @@ test_drpc_list_pools_success_with_pools(void **state)
 	D_FREE(resp.body.data);
 }
 
+/*
+ * dRPC List Containers setup/teardown
+ */
+
+static int
+drpc_list_cont_setup(void **state)
+{
+	mock_ds_mgmt_pool_list_cont_setup();
+
+	return 0;
+}
+
+static int
+drpc_list_cont_teardown(void **state)
+{
+	mock_ds_mgmt_pool_list_cont_teardown();
+
+	return 0;
+}
+
+/*
+ * dRPC List Containers tests
+ */
+static void
+pack_list_cont_req(Drpc__Call *call, Mgmt__ListContReq *req)
+{
+	size_t	len;
+	uint8_t	*body;
+
+	len = mgmt__list_cont_req__get_packed_size(req);
+	D_ALLOC(body, len);
+	assert_non_null(body);
+
+	mgmt__list_cont_req__pack(req, body);
+
+	call->body.data = body;
+	call->body.len = len;
+}
+
+static void
+setup_list_cont_drpc_call(Drpc__Call *call, char *uuid)
+{
+	Mgmt__ListContReq lc_req = MGMT__LIST_CONT_REQ__INIT;
+
+	lc_req.uuid = uuid;	/* pool UUID */
+	pack_list_cont_req(call, &lc_req);
+}
+
+static void
+expect_drpc_list_cont_resp_with_error(Drpc__Response *resp,
+				      int expected_err)
+{
+	Mgmt__ListContResp *lc_resp = NULL;
+
+	assert_int_equal(resp->status, DRPC__STATUS__SUCCESS);
+	assert_non_null(resp->body.data);
+
+	lc_resp = mgmt__list_cont_resp__unpack(NULL, resp->body.len,
+					      resp->body.data);
+	assert_non_null(lc_resp);
+
+	assert_int_equal(lc_resp->status, expected_err);
+
+	mgmt__list_cont_resp__free_unpacked(lc_resp, NULL);
+}
+static void
+expect_drpc_list_cont_resp_with_containers(Drpc__Response *resp,
+					   struct daos_pool_cont_info *cont,
+					   uint64_t cont_len)
+{
+	Mgmt__ListContResp	*cont_resp = NULL;
+	size_t			 i;
+
+	assert_int_equal(resp->status, DRPC__STATUS__SUCCESS);
+	assert_non_null(resp->body.data);
+
+	cont_resp = mgmt__list_cont_resp__unpack(NULL, resp->body.len,
+						 resp->body.data);
+	assert_non_null(cont_resp);
+	assert_int_equal(cont_resp->status, 0);
+
+	/* number of containers in response list == expected value. */
+	assert_int_equal(cont_resp->n_containers, cont_len);
+
+	/* number of containers in response list == total number in pool */
+	assert_int_equal(cont_resp->n_containers, cont_resp->numcontainers);
```

Comment:
Wondering about the purpose of the numcontainers param, since there is already an n_containers and they apparently contain the same value.

Reply:
From other discussion, RPC output should return the number of containers on failure, so we can realloc the right size. Not the situation in this particular test

Reply:
But that's in the CaRT RPC layer, right? I don't think there's any action for the control plane or the DMG tool to take if they get back the number of containers without any contents, since they didn't alloc a buffer in advance to hold the data.

Reply:
Yes, sorry. Had started out with this similar to the C API and also got a bit mixed up in the transition that happens when a drpc is handled for container list that also invokes CaRT RPC in the pool service.

I can simplify the protobuf definition to get rid of numcontainers.